Implement new connection establishment for SENDRECV and RDMA transports

rauteric opened the pull request:
Adds a new connection management module shared between the two transports.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

rauteric commented:
Finished some more of the implementation and split CM into multiple files, hopefully making it easier to comprehend.

The "missing features" in the PR description are still outstanding.

rauteric commented:
Added features. This is feature-complete now, except for locking. It also needs testing.

rauteric commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,107 @@
+	   since rx_req destructor returns buffer to freelist */
+	std::vector<std::unique_ptr<nccl_ofi_cm_rx_req>> rx_req_list;
+
+	/* rx reqs that need to be posted (due to EAGAIN) */
+	std::deque<nccl_ofi_cm_rx_req *> pending_rx_reqs;
```
I made this pending request list specific to the `rx_req` type, since the send requests don't necessarily need to be part of this list (they can be retried on the corresponding `test_ready()` functions.

An alternative is to make this a global pending reqs list, and add the send requests to this list as well.

bwbarrett reviewed (commented):
I didn't make it through most of the PR.  This doesn't match what we talked about in design.  I don't understand why we have CM comm structures.  To really review this properly needs some documentation that actually matches what the code is trying to do.  I'm frankly lost at what you're trying to do.

rauteric commented:
Added documentation mostly for the caller-facing (i.e., transport/protocol-facing) parts of the CM interface in `nccl_ofi_cm.h` and `nccl_ofi_cm_comms.h`.

rauteric commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,84 @@
+	nccl_ofi_cm_s_comm *connect(nccl_ofi_cm_handle *handle,
+				    const nccl_ofi_cm_ep_rail_info &rail_info);
+
+	nccl_ofi_cm_l_comm *get_l_comm(uint32_t l_comm_id);
```
These "comms" are just a temporary handle with address information from the remote side and a `->test_ready()` function for caller to test whether connection is complete.

I'm thinking maybe `{listen,send,recv}_connect_handle` is a better name?

rauteric replied:
As for the lookup table... it is needed when receiving a connect or connect response message, to look up the corresponding connect handle (the rx buffers are preposted to the endpoint and are not comm-specific [since we're not using tagged sends], so a lookup is necessary).

Since the rx buffers live in the `nccl_ofi_connect_manager` class, it seems the lookup table belongs in this class as well.

bwbarrett replied:
Then this doesn't feel like the right public interface for a CM object.  There's no reason code above this should ever need to do this translation.

rauteric commented on configure.ac:
```diff
@@ -192,6 +192,7 @@ AS_IF([test "${enable_picky_compiler}" != "no"], [
     dnl them. TODO: explicitly annotate those functions, rather than setting this
     dnl globally.
     ADD_PICKY_FLAGS([-Wno-unused-parameter])
+    ADD_PICKY_FLAGS([-Wno-invalid-offsetof])
```
I need `container_of`, which today the plugin implements using `offsetof`: https://github.com/aws/aws-ofi-nccl/blob/d809a88105a0e00d8101547488965e60a33bf59a/include/nccl_ofi_config_bottom.h#L33-L35

Looking around, I found some implementations of `container_of` which basically use a C++ version of `offsetof`, e.g., https://review.lttng.org/c/lttng-tools/+/8325/6. I doubt it works correctly in all cases (virtual inheritance seems to be particularly problematic), but it probably works in ours.

I'm not sure whether it's advantageous to use this instead of `offsetof` (which on gcc is implemented as a builtin). It would avoid this compiler warning to implement it ourselves, though.

rauteric commented on include/cm/nccl_ofi_cm_comms.h:
```diff
@@ -0,0 +1,189 @@
+	 */
+	nccl_ofi_cm_ep_rail_info get_sender_ep_rails();
+
+	/**
+	 * Set local rail addresses to be sent to the corresponding cm_s_comm in the
+	 * response message.
+	 */
+	void set_ep_rail_info(const nccl_ofi_cm_ep_rail_info &_ep_rail_info)
+	{
+		this->ep_rail_info = _ep_rail_info;
+	}
```
The plan here was to have the receive side choose the rails during `accept()`. This would allow choosing the rails depending on the sender's address, to support endpoint-per-comm mode.

During testing, I realized that this doesn't satisfy the requirement that `connect()` needs to complete without `accept()` being called.

To fix this, the CM needs to progress sending the connect response without `accept()` being called. I think the best way to do this is with a callback the transport provides to the CM to select the rails to be sent in the connect response. I will update the interface accordingly.

rauteric replied:
> During testing, I realized that this doesn't satisfy the requirement that `connect()` needs to complete without `accept()` being called.

We ended up revising this requirement, now we require connect() and accept() to be called simultaneously. Resolving.

rauteric commented:
Pushed an update to change the interface as suggested in [comment](https://github.com/aws/aws-ofi-nccl/pull/856#discussion_r2050878084).

Also added a stub of SENDRECV protocol integration. It passes the `nccl_connection` unit test, so the basic functionality seems working (data path not tested yet).

bwbarrett commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,228 @@
+ *       as well, and generally makes the code messy. May revisit this in the
+ *       future
+ */
+typedef nccl_ofi_cm_ep_rail_info (*select_rails_fn)(const nccl_ofi_cm_ep_rail_info &sender_rails,
```
I don't know why the rails concept is part of the connection management interface.

At some point, you're moving bits from place A to place B.  If that's all you're doing, I'm not really sure that a void* is a problem.  But the weird global scope function typedef, that I have questions about.

rauteric replied:
> I don't know why the rails concept is part of the connection management interface.

Agree. Discussed offline -- we will make this part of the connect/connect response messages opaque to the CM code -- transports will fill them in with rail addresses and anything else they need.

bwbarrett commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,228 @@
+{
+public:
+	/**
+	 * Initializes the CM system state. Creates an endpoint and posts
```
When are you imaging creating a CM object?  Every time a domain is created?

rauteric replied:
Yes, I think the CM object should live with the domain.

Today, I think the transports have a semantic that makes this complicated, in that the connection management functions (listen/connect/accept) are members of the endpoint, rather than the domain. I see two rational ways to refactor this:

1. (my preference) In each transport, move listen/connect/accept to the domain, rather than the endpoint, reserving transport endpoints specifically for data path communication (isend/irecv).

2. Keep the connection establishment APIs with the endpoint, and use this endpoint in the CM code, rather than creating a separate endpoint in the cm code. I dislike this option for similar reasons to [here](https://github.com/aws/aws-ofi-nccl/pull/856#discussion_r2055047222) -- I prefer the CM code to have ownership over a separate, non-transport-specific endpoint.

bwbarrett commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,228 @@
+	 * Initializes the CM system state. Creates an endpoint and posts
+	 * initial buffer pool
+	 *
+	 * @param cq: the completion queue to bind the new endpoint to.
```
you didn't document much of the interface.  Why do we need a reference to the mr key pool?  Why should the interface take a bunch of pointers to Libfabric objects?  Should it just take a reference to a transport domain object?

rauteric replied:
Added documentation to these parameters, but clarifying here:

**Connection Manager Endpoint Creation**

1. The CM creates its own endpoint rather than creating a transport-specific endpoint
2. This requires both `fi_info` and `fi_domain` objects for the `fi_endpoint` call

*Alternative approach (decided against): Using `nccl_ofi_domain_t::get_ep()` to create transport-specific endpoints for use by the connect manager.
This was avoided because transport-specific endpoints include extra transport-specific behaviors unnecessary for the CM. Especially the RDMA protocol, which creates endpoints on each rail and posts RDMA-specific rx buffers.*

**Memory Registration Key Management**

The CM handles its own memory registration.

For `FI_MR_PROV_KEY` mode, CM must use keys distinct from transport-selected keys. Hence the need to use the same mr key pool (now passed as a reference rather than a pointer)

bwbarrett commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,228 @@
+	 * Return a communicator (TODO rename to handle or some such) that can
+	 * be used to accept incoming connections from connect()
+	 */
+	nccl_ofi_cm_l_comm *listen();
```
Why is this a pointer instead of either an object or a reference?  We're not writing C code any more, so let's act like it.

rauteric replied:
> Why is this a pointer instead of either an object or a reference? We're not writing C code any more, so let's act like it.

Ack. Was thinking about this particular API (`listen()`) -- I don't think a reference makes sense, because it is returning a dynamically allocated object which needs to be freed, and an object doesn't make sense, because a listener shouldn't be copied. I decided on a unique pointer.

bwbarrett commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,228 @@
+
+
+	/**
+	 * Provide a "rail selector" used to choose rails to advertise
```
Again, why all the rail crap in the CM?  That should be totally opaque to the CM.

bwbarrett commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,84 @@
+	nccl_ofi_cm_l_comm *get_l_comm(uint32_t l_comm_id);
+	nccl_ofi_cm_s_comm *get_s_comm(uint32_t s_comm_id);
+
+	nccl_ofi_freelist_elem_t *alloc_conn_msg()
```
Why is the right thing to return from this (PUBLIC) API call a pointer to a freelist elem, thereby destroying any type safety?

rauteric replied:
> Why is the right thing to return from this (PUBLIC) API call a pointer to a freelist elem, thereby destroying any type safety?

This is no longer in the public API. However, it still is a pointer to a freelist elem (so that we get a structure containing the data pointer and the memory registration together for free).

bwbarrett commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,84 @@
+		nccl_ofi_freelist_entry_free(conn_msg_fl, conn_msg);
+	}
+
+	fid_ep *get_ep() {return ep;}
```
Also, again don't understand ownership here.

bwbarrett commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,228 @@
+
+	const cm_ep_name &get_conn_ep_name() {return conn_ep_name;}
+
+	nccl_ofi_idpool_t *get_l_comm_id_pool() { return &l_comm_id_pool; }
```
Why pointers?

bwbarrett commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,228 @@
+	nccl_ofi_idpool_t *get_l_comm_id_pool() { return &l_comm_id_pool; }
+	nccl_ofi_idpool_t *get_data_comm_id_pool() { return &data_comm_id_pool; }
+
+	std::unordered_map<uint32_t, nccl_ofi_cm_l_comm *> *get_l_comm_map()
```
Let's say tomorrow you decide that you need the connection lookup to be linear instead of logarithmic.  So you change the type of the conn map to be a vector.  Now you've got to change this interface function as well, which just makes no sense from an abstraction point of view.

bwbarrett commented on include/cm/nccl_ofi_cm_comms.h:
```diff
@@ -0,0 +1,187 @@
+	/**
+	 * Construct a new cm_l_comm
+	 *
+	 * Not to be called by the transport code -- the transport code
```
Some `protected` and `friend` keywords seem way better than a comment here.  And again with the pointers.

bwbarrett commented on include/cm/nccl_ofi_cm_comms.h:
```diff
@@ -0,0 +1,187 @@
+	/**
+	 * Obtain the handle associated with this cm_l_comm
+	 *
+	 * The caller (transport) should return this handle back to NCCL to
```
tell me more?

bwbarrett commented on include/cm/nccl_ofi_cm_comms.h:
```diff
@@ -0,0 +1,187 @@
+	nccl_ofi_cm_l_comm(nccl_ofi_connection_manager *cm);
+	~nccl_ofi_cm_l_comm();
+
+	int process_conn_msg(const nccl_ofi_cm_conn_msg &conn_msg);
```
I'm not sure what this does, but pretty sure it's a problem with the interface.

rauteric replied:
> I'm not sure what this does, but pretty sure it's a problem with the interface.

This still exists, but no longer in the public API.

bwbarrett commented on include/cm/nccl_ofi_cm_comms.h:
```diff
@@ -0,0 +1,187 @@
+
+	int process_conn_msg(const nccl_ofi_cm_conn_msg &conn_msg);
+
+	void insert_completed_r_comm(nccl_ofi_cm_r_comm *r_comm)
```
ugh

rauteric commented:
The latest commit I pushed attempts to make a more sensible public interface in `nccl_ofi_cm.h`, clearly separating transport-usable functionality from internal implementation details. Corresponding implementation is not done yet.

bwbarrett commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,317 @@
+	 * Note: Returns a pointer to memory owned by this object. The memory
+	 * is valid until this object is destroyed.
+	 */
+	const void *get_conn_msg()
```
it's not really a connection message, is it?  It's really data you're getting.

bwbarrett commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,317 @@
+	 * Return opaque output pointer provided by transport from
+	 * transport_select_conn_resp_fn for this connection
+	 */
+	void *get_transport_select_output()
```
i honestly have no idea what this does.

rauteric replied:
Resolved in https://github.com/aws/aws-ofi-nccl/pull/856#discussion_r2056787768, this won't be needed.

rauteric commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,317 @@
+ *      For example, this might be a pointer to the transport endpoint created
+ *      by this function.
+ */
+typedef void (*transport_select_conn_resp_fn)(const void *transport_connect_msg,
```
Discussed offline; will make this a `nccl_ofi_domain_t::accept_connection()` callback that returns a `nccl_ofi_comm_t` object. The latter object will have a `nccl_ofi_comm_t::complete_connection()` callback for when the connection is established. This gets rid of much of the "opaque crap" (we'll only have the transport-specific content of the connect [resp] messages left).

rauteric commented:
Updated public API for revised requirement that `connect()` and `accept()` are progressed simultaneously. In this case we don't need the transport callback to set the connect response message.

rauteric commented:
Force-push to add tie-in of new CM interface to RDMA protocol.

rauteric commented:
Latest force-push reworks the internal implementation of the CM to match the new public interface.

I reverted changes to SENDRECV protocol until they can be reworked with the new interface. RDMA protocol tie-in to new interface is nearly complete -- just need to refactor the completion queue so that it is shared across the domain.

rauteric commented on src/cm/nccl_ofi_cm_reqs.cpp:
```diff
@@ -0,0 +1,215 @@
+	done_callback(_done_callback)
+{
+	/**
+	 * We use a different behavior depending on the value of
+	 * data_progress_auto
```
Calling this out for review. If supported by the provider, `fi_inject` is the best way to handle the multi-recv requirement when the provider supports `FI_PROGRESS_AUTO`.

Depending on provider support, we may also want to support providers with `FI_PROGRESS_AUTO` but no inject.

rauteric replied:
> Calling this out for review. If supported by the provider, `fi_inject` is the best way to handle the multi-recv requirement when the provider supports `FI_PROGRESS_AUTO`.

Upon testing with Libfabric 2.1.0, this doesn't work with efa-direct, which has a max inject size of 32.

Will rework this to not use inject. Shouldn't be a problem -- basically, we keep the request alive until we get the completion, but still return from accept() immediately.

rauteric replied:
Fixed -- replaced inject with a strategy of `receiver->test_ready` returning true immediately after sending the conn response, and cleaning up the request later when we get the delivery completion.

rauteric commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,256 @@
+
+typedef char ep_name[MAX_EP_ADDR];
+
+struct cm_ep_name {
```
This has been renamed to `cm_ep_addr`.

Also, most of the internal bits of the CM code are now in `nccl_ofi_cm` namespace.

rauteric commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,84 @@
+	fid_ep *get_ep() {return ep;}
+	fid_domain *get_domain() {return domain;}
+
+	int av_insert_address(ep_name address, fi_addr_t *fi_addr);
```
Moved to an interface for the (private) endpoint class.

rauteric commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,317 @@
+	 *
+	 * See documentation below on nccl_ofi_cm_receiver_info
+	 */
+	std::optional<nccl_ofi_cm_receiver_info> accept();
```
> Why is a polling accept() interface the right path vs. an iaccept() / callback or iaccept() / test style interface? I'd push for iaccept that takes an rcomm reference and calls a function on that rcomm when the data is available.

Talked offline -- I believe we need a polling approach to deal with EAGAIN retries.

Agreed that polling is not ideal -- will revisit when we think of a better way.

bwbarrett reviewed (commented):
Didn't make it very far, but ran out of time.  Leaving the wip here.

rauteric commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,277 @@
+	 *
+	 * @param data: transport-provided buffer of size conn_msg_size
+	 */
+	void set_conn_resp_msg_data(const void *data);
```
> how do you not need a length here?

The length of the transport-specific connect message data is passed to the `nccl_ofi_connection_manager` constructor. We need it there so that the CM can post receive buffers of appropriate size. I'm using that length throughout the code instead of passing it into functions like these.

I'm not sure passing the size again to this function (`set_conn_resp_msg_data`) would serve much purpose, besides maybe an assert that the size passed here matches the size passed to `nccl_ofi_connection_manager` constructor.

rauteric commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,277 @@
+	 *      size of transport-specific part of connect and connect response
+	 *      messages
+	 */
+	nccl_ofi_connection_manager(fi_info *info, fid_domain *domain, fid_cq *cq,
```
> I think I left this comment already, but especially in light of the failure mode discussions we're having, should this be a plugin domain object instead of individual fields? If the domain closes the CQ, you have no way of knowing in this path.

Yeah, we can pass in a domain object here to consolidate arguments and as a sanity check that the domain is not closed. (We'll need to add a new getter to `nccl_net_ofi_domain_t` to get the info and domain objects.)

However, I think it should be the transport's responsibility to check that the domain/CQ are still open before calling into the CM.

rauteric replied:
Done: [d287c7a](https://github.com/aws/aws-ofi-nccl/pull/856/commits/d287c7a1461f60febf684909abae15a18890b71e)

rauteric commented on include/cm/nccl_ofi_cm_resources.h:
```diff
@@ -0,0 +1,317 @@
+};
+
+/**
+ * Map from ID to references to the templated parameter.
```
> Really not sure why this class, instead of just using an unordered_map?

I don't have a strong opinion, but I chose to use a wrapper class because:
1. The `unordered_map` interface is a bit clunky in places, especially `map.find()` returning an iterator
2. This way we can change the implementation, say to use a vector instead of `unordered_map`, without changing the rest of the code (I thought that's the point you were making in [comment](https://github.com/aws/aws-ofi-nccl/pull/856#discussion_r2052828549)).

rauteric commented:
Force-push to [500aa3d](https://github.com/aws/aws-ofi-nccl/commit/500aa3d7632d0337d1e9fa03f5cfb84ca82bce15) is a rebase on master.

rauteric commented:
Summary of recent commits:
- [39e706f](https://github.com/aws/aws-ofi-nccl/pull/856/commits/39e706fde912810306d789655c48f61659a8cc2c): Move test_ready to return a single int for status and error codes
- [e54f618](https://github.com/aws/aws-ofi-nccl/pull/856/commits/e54f6188ef99278a43b27ee54f254f3b0168e643): SENDRECV tie-in
- [c40658c](https://github.com/aws/aws-ofi-nccl/pull/856/commits/c40658c41e5543dc63d5a38df3a40cb8838e302f): updates transport `accept()` routines to support multiple accepts per listen (needs testing since NCCL doesn't do this today)
- [6686920](https://github.com/aws/aws-ofi-nccl/pull/856/commits/6686920be5b02d81b7f2793fce8433dd313d60ca): Locking
- [716a7ce](https://github.com/aws/aws-ofi-nccl/pull/856/commits/716a7cec5af67642379ef08687b872adb9281f26): Parameterize number of CM RX buffers to post

---

At this point, as far as I know, this PR is feature-complete -- tested with functional tests and NCCL tests. I'm still addressing feedback.

rauteric commented:
Commit [d287c7a](https://github.com/aws/aws-ofi-nccl/pull/856/commits/d287c7a1461f60febf684909abae15a18890b71e) refactors CM to take `nccl_net_ofi_domain_t` object as a parameter instead of individual OFI objects.

rauteric commented:
Pushed commits:
* [530e6a1](https://github.com/aws/aws-ofi-nccl/pull/856/commits/530e6a11aac00ad95487826b6d38ed484905df10): Map to callbacks instead of listener/connectors (and thus avoid having rx_req as a friend class)
* [269f2a9](https://github.com/aws/aws-ofi-nccl/pull/856/commits/269f2a90d1497df8e0f4945993afe90b2527c0d2): bool instead of std::optional
* [21d5f4d](https://github.com/aws/aws-ofi-nccl/pull/856/commits/21d5f4d49809a0823ef8b8a2f36b2d67e69987cd): endpoint::send/recv API fix

rauteric commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,277 @@
+	std::optional<std::vector<uint8_t>> conn_resp_msg_data;
+	nccl_ofi_cm::nccl_ofi_cm_send_conn_req *send_conn_req;
+
+	bool conn_msg_sent;
```
Turned these into a bitfield (assuming you ment bit field, not bit mask): [269f2a9](https://github.com/aws/aws-ofi-nccl/pull/856/commits/269f2a90d1497df8e0f4945993afe90b2527c0d2)

rauteric commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,277 @@
+	/* For constructor */
+	friend class nccl_ofi_connection_manager;
+	/* For process_conn_msg */
+	friend class nccl_ofi_cm::nccl_ofi_cm_rx_req;
```
Resolved the `nccl_ofi_cm::nccl_ofi_cm_rx_req` friend class by using callbacks instead, which is an overall cleaner approach. [530e6a1](https://github.com/aws/aws-ofi-nccl/pull/856/commits/530e6a11aac00ad95487826b6d38ed484905df10)

rauteric commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,277 @@
+	 * pointer to memory owned by this object, and is invalidated when
+	 * send_connector is destroyed.
+	 */
+	const void *get_conn_resp_msg();
```
Same answer in https://github.com/aws/aws-ofi-nccl/pull/856#discussion_r2069438541.

rauteric commented:
Marked this PR ready for review. The commit history is unwieldy (partly to aid review), but will squash into something sensible before commit.

bwbarrett commented on src/nccl_ofi_sendrecv.cpp:
```diff
@@ -1520,6 +1380,34 @@ static nccl_net_ofi_sendrecv_recv_comm_t *sendrecv_recv_comm_prepare(nccl_net_of
 }
 
+
+static inline char *sendrecv_get_local_address(struct fid_ep *ep);
```
isn't this a uint8_t array?  It's not a string.

bwbarrett commented on src/nccl_ofi_sendrecv.cpp:
```diff
@@ -1563,10 +1444,10 @@ static int sendrecv_listen_comm_accept(nccl_net_ofi_listen_comm_t *listen_comm,
+	nccl_ofi_connection_info_t conn_resp_msg = { };
 	
 	/*
 	 * Take appropriate actions based on connection stage of communicator.
```
Part of the goal of this work was to get rid of the huge error prone switch statements below.  We didn't make progress on that, which is disappointing.  That should all basically be event driven, not polling driven.

rauteric replied:
In upcoming revision I have made some progress on this. I entirely removed the states on the connect side; on the accept side, I reduced the number of states.

I couldn't entirely get rid of the states for `accept` in part because of the multi-recv requirement (documented in [comment](https://github.com/rauteric/aws-ofi-nccl/blob/21d5f4d49809a0823ef8b8a2f36b2d67e69987cd/src/cm/nccl_ofi_cm_reqs.cpp#L141-L146)) that we only send the connect response message when we return an `r_comm` from `accept()`. So the sending of connect response can't be cq event driven; it has to happen inside `accept()`.

bwbarrett commented on include/nccl_ofi_param.h:
```diff
@@ -427,4 +427,13 @@ OFI_NCCL_PARAM_INT(force_num_rails, "FORCE_NUM_RAILS", 0);
+ * Posting buffers will use more memory, but may make connection establishment
+ * complete more quickly, especially with large numbers of ranks.
+ */
+OFI_NCCL_PARAM_UINT(cm_num_rx_buffers, "CM_NUM_RX_BUFFERS", 32);
```
How are you going to find the right number for this?

bwbarrett reviewed (commented):
This is really hard to review because you haven't properly squashed your commits, and a good chunk of your commits have changes later.  But here's some general comments.

rauteric commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,275 @@
+	std::vector<uint8_t> conn_resp_msg_data;
+	nccl_ofi_cm::nccl_ofi_cm_send_conn_req *send_conn_req;
+
+	bool conn_msg_sent : 1;
```
Assuming you mean why the bitfield... I thought you were asking for that in https://github.com/aws/aws-ofi-nccl/pull/856#discussion_r2067471810.

It does make the `nccl_ofi_cm_send_connector` class a bit smaller, but it's probably not important enough to be useful. I will remove it.

rauteric commented:
Addressed feedback, and properly squashed this PR into two commits (adding the new connection management code, and implementing it in both transports).

hershys-aws commented on include/cm/nccl_ofi_cm.h:
```diff
@@ -0,0 +1,280 @@
+	 * Note: the caller takes ownership of the memory associated with this
+	 * object and should release it by deleting the pointer.
+	 */
+	nccl_ofi_cm_receiver *accept();
```
Could be a future improvement given time: I know we are still in progress of moving from C->C++ but thoughts on wrapping this in a shared_pointer or even custom lightweight implementation so we don't need the onus of memory management on the caller?

rauteric replied:
Two things discussed here were 1) returning a unique pointer or 2) returning a lightweight handle that will destroy the receiver when it goes out of scope (which is basically reimplementing unique ptr).

Neither approach will buy us anything until we convert the send/recv comms to C++ classes, so we get automatic out-of-scoping. After that we should revisit the raw pointers here.

hershys-aws commented on include/cm/nccl_ofi_cm_resources.h:
```diff
@@ -0,0 +1,245 @@
+ */
+class cm_resources
+{
+public:
+	/* Mutex guarding all CM operations */
+	std::mutex cm_mutex;
+	/* Endpoint for CM operations */
+	endpoint ep;
+private:
+	size_t conn_msg_data_size;
+
+public:
```
Why keep switching? Just combine it with the public and private below...

rauteric replied:
> Why keep switching? Just combine it with the public and private below...

Ok, after seeing the CI failure, I realized why I had done this. The class members are initialized in order, and `conn_msg_data_size` needs to be initialized before `buff_mgr`, which depends on it.

I reverted to this public/private switching, and put an explanatory comment.

hershys-aws reviewed (commented):
Just added questions/comments.

rauteric commented:
Addressed feedback from Hershel.

rauteric commented:
Resolved build issue.

rauteric commented:
Rebased on master and attempted to resolve CI issue.

rauteric commented:
[Force-push](https://github.com/aws/aws-ofi-nccl/compare/cc8d7272bd88a5f9df79386fbcc95e2605d0ce28..282e1cfdbe170f54081f1d94807df8f5e9bb4ae1) fixes a bug in the RDMA connect path.

I haven't addressed the recent feedback on this PR yet.

xwangxin commented on src/nccl_ofi_rdma.cpp:
```diff
@@ -5197,16 +4943,21 @@ static int accept(nccl_net_ofi_listen_comm_t *listen_comm,
 			return 0;
 		}
 
+		{
```
Why do you need to {} here?

rauteric replied:
Yeah, the large case statements (and large functions in general) are problematic.

In the latest commit ([2f3e648](https://github.com/aws/aws-ofi-nccl/commit/2f3e6483f4efb44a6bff8f6ab7d881edd4ffc58a)) I tried to make this situation better by refactoring a chunk of `accept()` into a function.

rauteric commented on src/cm/nccl_ofi_cm.cpp:
```diff
@@ -0,0 +1,290 @@
+	return ret;
+}
+
+void nccl_ofi_cm_receiver::set_conn_resp_msg_delivered()
+{
+	conn_resp_msg_delivered = true;
+	/* conn_resp_req will delete itself, so unset the pointer */
+	conn_resp_req = nullptr;
+}
```
You mean just put these two lines in the lambda, and get rid of this function? Yeah, that's probably better.

hershys-aws replied:
Yeah, there was a couple other places with similar behavior you can probably use the sample thing.

rauteric commented on src/cm/nccl_ofi_cm.cpp:
```diff
@@ -0,0 +1,290 @@
+
+	/* Copy user data after conn resp msg */
+	assert(size == resources.get_conn_msg_data_size());
+	memcpy(&conn_resp_msg + 1, data, size);
```
This is to copy the transport-specific data to the memory immediately following the CM-specific data.

Like Hershel [suggested](https://github.com/aws/aws-ofi-nccl/pull/856#discussion_r2124995395), this should really be a macro with explanation.

rauteric commented on src/cm/nccl_ofi_cm.cpp:
```diff
@@ -0,0 +1,290 @@
+nccl_ofi_cm_receiver *nccl_ofi_cm_listener::accept()
+{
+	std::lock_guard<std::mutex> lock(resources.cm_mutex);
+	int ret = resources.pending_reqs_queue.process_pending_reqs();
```
We need it in `accept()` to ensure we have enough rx buffers available to receive the connect message. (If for some reason all rx buffers ended up in pending queue, we will never get to calling `test_ready`, so we will never repost these unless we do it during `accept()`.)

We need it in `test_ready()` in case we hit EAGAIN while sending the connect response message.

rauteric commented on include/cm/nccl_ofi_cm_resources.h:
```diff
@@ -0,0 +1,250 @@
+	/* Endpoint for CM operations */
+	endpoint ep;
+
+private:
```
This was brought up in https://github.com/aws/aws-ofi-nccl/pull/856#discussion_r2114485636; the problem is that these fields need to be initialized in order. Since `buff_mgr` (used to allocate connect messages from freelist) depends on `conn_msg_data_size` (influences the size of the buffers), the latter needs to be initialized first.

An alternative would be to make members like `buff_mgr` private and have a getter (something like `get_buff_mgr()`), I felt this approach is overly verbose.

rauteric commented on include/cm/nccl_ofi_cm_resources.h:
```diff
@@ -0,0 +1,250 @@
+	 * Return the next available connector ID. Used for listeners and
+	 * send_connectors.
+	 */
+	uint64_t get_next_connector_id() { return next_connector_id++; }
```
The entire CM code is protected by a mutex (`cm_resources::cm_mutex`), making further atomics unnecessary.

Now that the transports take a domain lock within the connection establishment routines, even the `cm_mutex` is now unnecessary. I will consider removing it at some point.

rauteric commented on include/cm/nccl_ofi_cm_resources.h:
```diff
@@ -0,0 +1,250 @@
+private:
+	/* Size of the transport-specific data in the connect messages This must
+	   be initialized before get_conn_msg{_data}_size() is called. */
+	size_t conn_msg_data_size;
```
Yes, for a given transport, the size of the connect message data is known at compile time. But since the CM code is shared between transports, the size needs to be passed as a parameter to the CM constructor.

It could also be made a template parameter, but that significantly increases the verbosity of the code (since all the CM classes will need the template parameter), for not much gain.

rauteric commented:
[Force-pushes](https://github.com/aws/aws-ofi-nccl/compare/282e1cfdbe170f54081f1d94807df8f5e9bb4ae1..8de27efb02d0596dc31851f024ac3f2b2b8ba150) from [282e1cf](https://github.com/aws/aws-ofi-nccl/commit/282e1cfdbe170f54081f1d94807df8f5e9bb4ae1) to [8de27ef](https://github.com/aws/aws-ofi-nccl/commit/8de27efb02d0596dc31851f024ac3f2b2b8ba150) address feedback and fix bugs.

rauteric commented:
Force-push ^ is a rebase on master.
